jar-infer/jar-infer-lib/src/main/java/com/uber/nullaway/jarinfer/DefinitelyDerefedParams.java

```diff
@@ -98,15 +104,60 @@ private static void LOG(boolean cond, String tag, String msg) {
   Set<Integer> analyze() {
     // Get ExceptionPrunedCFG
     LOG(DEBUG, "DEBUG", "@ " + method.getSignature());
-    Set<Integer> derefedParamList = new HashSet<Integer>();
     prunedCFG = ExceptionPrunedCFG.make(cfg);
     // In case the only control flows are exceptional, simply return.
     if (prunedCFG.getNumberOfNodes() == 2
         && prunedCFG.containsNode(cfg.entry())
         && prunedCFG.containsNode(cfg.exit())
         && GraphUtil.countEdges(prunedCFG) == 0) {
-      return derefedParamList;
+      return new HashSet<>();
     }
+
+    // Get number of params and value number of first param
+    int numParam = ir.getSymbolTable().getNumberOfParameters();
+    int firstParamIndex =
+        method.isStatic() ? 1 : 2; // 1-indexed; v1 is 'this' for non-static methods
+
+    Set<Integer> derefedParamList;
+    if (USE_EXTENDED_APPROACH) {
+      derefedParamList = computeDerefParamList(numParam, firstParamIndex);
+    } else {
+      derefedParamList = computeDerefParamListUsingPDom(numParam, firstParamIndex);
+    }
+    return derefedParamList;
+  }
+
+  @NotNull
+  private Set<Integer> computeDerefParamList(int numParam, int firstParamIndex) {
+    Set<Integer> derefedParamList = new HashSet<>();
+    Map<ISSABasicBlock, Set<Integer>> blockToDerefSetMap = new HashMap<>();
+
+    // find which params are definitely non-null in each block
+    prunedCFG.forEach(
+        basicBlock -> {
+          Set<Integer> derefParamSet = new HashSet<>();
+          blockToDerefSetMap.put(basicBlock, derefParamSet);
+          checkForUseOfParams(derefParamSet, numParam, firstParamIndex, basicBlock);
+        });
+
+    // for each param check if there is a path from entry to exit going through basic blocks
+    // which do not guarantee that the param is non-null
```

Comment:
So, not sure this comment is right, the way I understand the code below, it says:

"For each param `p`, if no path exists from the entry block to the exit block which traverses only basic blocks which do not dereference  `p`, then mark `p` as `@NonNull`"

Reply:
Other than the comment being confusing, this approach actually looks broadly correct to me, with the only issues being: 

a) Code sets the argument `p` when it originally is `null` such as:
```
public void test(String s, String t) {
   Objects.requireNonNull(s);
   if (t == null) {
     t = s;
   }
   t.toString();
}
```
(It's probably fine that we don't handle this, but worth checking if this produces a lot of new false positives for monorepo)

b) The whole other issue about exceptional control flow being pruned that might start to matter the more stuff our analysis catches otherwise.

Also, probably worth checking the performance impact of this vs the original approach. I don't think it should be too bad, since all the time probably went into building the IR and GFG either way, but I think it's worth keeping track.

Reply:
a) works because of SSA, since variables are assigned only one time. A new variable(v8) is created 

```
public void test(String s, String t) {
   Objects.requireNonNull(s);
   if (t == null) {
     t = s;
   }
   Objects.requireNonNull(t);
}

13:41:12.302 [DEBUG] [TestEventLogger]                v8 = phi  v3,v2
13:41:12.303 [DEBUG] [TestEventLogger]     9   v10 = invokestatic < Application, Ljava/util/Objects, requireNonNull(Ljava/lang/Object;)Ljava/lang/Object; > v8 @12 exception:v9(line 17)
```

Reply:
Oh, right, that makes sense. Nice! Could we add that example as a test then? :)

Reply:
Yes.. adding.